// @azure/container-registry/dist-esm/src/generated/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ApiVersion20210701} that the service accepts. */
export var KnownApiVersion20210701;
(function (KnownApiVersion20210701) {
    /** Api Version '2021-07-01' */
    KnownApiVersion20210701["TwoThousandTwentyOne0701"] = "2021-07-01";
})(KnownApiVersion20210701 || (KnownApiVersion20210701 = {}));
/** Known values of {@link ArtifactArchitecture} that the service accepts. */
export var KnownArtifactArchitecture;
(function (KnownArtifactArchitecture) {
    /** i386 */
    KnownArtifactArchitecture["I386"] = "386";
    /** AMD64 */
    KnownArtifactArchitecture["Amd64"] = "amd64";
    /** ARM */
    KnownArtifactArchitecture["Arm"] = "arm";
    /** ARM64 */
    KnownArtifactArchitecture["Arm64"] = "arm64";
    /** MIPS */
    KnownArtifactArchitecture["Mips"] = "mips";
    /** MIPSLE */
    KnownArtifactArchitecture["MipsLe"] = "mipsle";
    /** MIPS64 */
    KnownArtifactArchitecture["Mips64"] = "mips64";
    /** MIPS64LE */
    KnownArtifactArchitecture["Mips64Le"] = "mips64le";
    /** PPC64 */
    KnownArtifactArchitecture["Ppc64"] = "ppc64";
    /** PPC64LE */
    KnownArtifactArchitecture["Ppc64Le"] = "ppc64le";
    /** RISCv64 */
    KnownArtifactArchitecture["RiscV64"] = "riscv64";
    /** s390x */
    KnownArtifactArchitecture["S390X"] = "s390x";
    /** Wasm */
    KnownArtifactArchitecture["Wasm"] = "wasm";
})(KnownArtifactArchitecture || (KnownArtifactArchitecture = {}));
/** Known values of {@link ArtifactOperatingSystem} that the service accepts. */
export var KnownArtifactOperatingSystem;
(function (KnownArtifactOperatingSystem) {
    /** Aix */
    KnownArtifactOperatingSystem["Aix"] = "aix";
    /** Android */
    KnownArtifactOperatingSystem["Android"] = "android";
    /** Darwin */
    KnownArtifactOperatingSystem["Darwin"] = "darwin";
    /** Dragonfly */
    KnownArtifactOperatingSystem["Dragonfly"] = "dragonfly";
    /** FreeBsd */
    KnownArtifactOperatingSystem["FreeBsd"] = "freebsd";
    /** Illumos */
    KnownArtifactOperatingSystem["Illumos"] = "illumos";
    /** IOS */
    KnownArtifactOperatingSystem["IOS"] = "ios";
    /** JS */
    KnownArtifactOperatingSystem["JS"] = "js";
    /** Linux */
    KnownArtifactOperatingSystem["Linux"] = "linux";
    /** NetBsd */
    KnownArtifactOperatingSystem["NetBsd"] = "netbsd";
    /** OpenBsd */
    KnownArtifactOperatingSystem["OpenBsd"] = "openbsd";
    /** Plan9 */
    KnownArtifactOperatingSystem["Plan9"] = "plan9";
    /** Solaris */
    KnownArtifactOperatingSystem["Solaris"] = "solaris";
    /** Windows */
    KnownArtifactOperatingSystem["Windows"] = "windows";
})(KnownArtifactOperatingSystem || (KnownArtifactOperatingSystem = {}));
/** Known values of {@link PostContentSchemaGrantType} that the service accepts. */
export var KnownPostContentSchemaGrantType;
(function (KnownPostContentSchemaGrantType) {
    /** AccessTokenRefreshToken */
    KnownPostContentSchemaGrantType["AccessTokenRefreshToken"] = "access_token_refresh_token";
    /** AccessToken */
    KnownPostContentSchemaGrantType["AccessToken"] = "access_token";
    /** RefreshToken */
    KnownPostContentSchemaGrantType["RefreshToken"] = "refresh_token";
})(KnownPostContentSchemaGrantType || (KnownPostContentSchemaGrantType = {}));
